File: sidemantic/core/inheritance.py
"""Inheritance utilities for models and metrics."""

from sidemantic.core.dimension import Dimension
from sidemantic.core.metric import Metric
from sidemantic.core.model import Model
from sidemantic.core.relationship import Relationship
from sidemantic.core.segment import Segment


def merge_model(child: Model, parent: Model) -> Model:
    """Merge child model with parent model using inheritance.

    Child model inherits all fields from parent, with child values taking precedence.

    Args:
        child: Child model with extends reference
        parent: Parent model to inherit from

    Returns:
        New model with merged properties

    Examples:
        >>> parent = Model(name="base", table="base_table", dimensions=[...])
        >>> child = Model(name="child", extends="base", dimensions=[...])
        >>> merged = merge_model(child, parent)
    """
    # Start with parent's data
    merged_data = parent.model_dump(exclude={"name"})

    # Override with child's data (excluding None values and extends)
    child_data = child.model_dump(exclude_none=True, exclude={"extends"})

    # Merge lists (dimensions, metrics, relationships, segments)
    # Child's items are added to parent's items
    for field in ["dimensions", "metrics", "relationships", "segments"]:
        parent_items = merged_data.get(field, [])
        child_items = child_data.get(field, [])

        # Combine, with child items overriding parent items with same name
        parent_by_name = {item["name"]: item for item in parent_items}
        child_by_name = {item["name"]: item for item in child_items}

        # Merge
        parent_by_name.update(child_by_name)
        merged_data[field] = list(parent_by_name.values())

    # Override scalar fields with child values
    for field in ["table", "sql", "description", "primary_key"]:
        if field in child_data:
            merged_data[field] = child_data[field]

    # Keep child's name
    merged_data["name"] = child.name

    # Reconstruct model objects from dicts
    if merged_data.get("dimensions"):
        merged_data["dimensions"] = [Dimension(**d) for d in merged_data["dimensions"]]
    if merged_data.get("metrics"):
        merged_data["metrics"] = [Metric(**m) for m in merged_data["metrics"]]
    if merged_data.get("relationships"):
        merged_data["relationships"] = [Relationship(**r) for r in merged_data["relationships"]]
    if merged_data.get("segments"):
        merged_data["segments"] = [Segment(**s) for s in merged_data["segments"]]

    return Model(**merged_data)


def merge_metric(child: Metric, parent: Metric) -> Metric:
    """Merge child metric with parent metric using inheritance.

    Child metric inherits all fields from parent, with child values taking precedence.

    Args:
        child: Child metric with extends reference
        parent: Parent metric to inherit from

    Returns:
        New metric with merged properties

    Examples:
        >>> parent = Metric(name="base_revenue", agg="sum", sql="amount")
        >>> child = Metric(name="filtered_revenue", extends="base_revenue", filters=["status = 'completed'"])
        >>> merged = merge_metric(child, parent)
    """
    # Start with parent's data
    merged_data = parent.model_dump(exclude={"name"})

    # Override with child's data (excluding None values and extends)
    child_data = child.model_dump(exclude_none=True, exclude={"extends"})

    # Handle list fields - merge arrays
    for field in ["filters", "drill_fields"]:
        parent_items = merged_data.get(field) or []
        child_items = child_data.get(field)

        if child_items:
            # Child can add to parent's list
            merged_data[field] = parent_items + child_items
        else:
            merged_data[field] = parent_items if parent_items else None

    # Override all other fields with child values
    for field, value in child_data.items():
        if field not in ["filters", "drill_fields", "extends", "name"]:
            merged_data[field] = value

    # Keep child's name
    merged_data["name"] = child.name

    return Metric(**merged_data)


def resolve_model_inheritance(models: dict[str, Model]) -> dict[str, Model]:
    """Resolve inheritance for all models in a graph.

    Args:
        models: Dictionary of model name to model

    Returns:
        Dictionary of model name to resolved model (with inheritance applied)

    Raises:
        ValueError: If circular inheritance detected or parent not found
    """
    resolved = {}
    in_progress = set()

    def resolve(name: str) -> Model:
        if name in resolved:
            return resolved[name]

        if name in in_progress:
            raise ValueError(f"Circular inheritance detected for model '{name}'")

        model = models.get(name)
        if not model:
            raise ValueError(f"Model '{name}' not found")

        # If no inheritance, just return as-is
        if not model.extends:
            resolved[name] = model
            return model

        # Resolve parent first
        in_progress.add(name)
        parent = resolve(model.extends)
        in_progress.remove(name)

        # Merge child with parent
        merged = merge_model(model, parent)
        resolved[name] = merged
        return merged

    # Resolve all models
    for name in models:
        resolve(name)

    return resolved


def resolve_metric_inheritance(metrics: dict[str, Metric]) -> dict[str, Metric]:
    """Resolve inheritance for all metrics in a graph.

    Args:
        metrics: Dictionary of metric name to metric

    Returns:
        Dictionary of metric name to resolved metric (with inheritance applied)

    Raises:
        ValueError: If circular inheritance detected or parent not found
    """
    resolved = {}
    in_progress = set()

    def resolve(name: str) -> Metric:
        if name in resolved:
            return resolved[name]

        if name in in_progress:
            raise ValueError(f"Circular inheritance detected for metric '{name}'")

        metric = metrics.get(name)
        if not metric:
            raise ValueError(f"Metric '{name}' not found")

        # If no inheritance, just return as-is
        if not metric.extends:
            resolved[name] = metric
            return metric

        # Resolve parent first
        in_progress.add(name)
        parent = resolve(metric.extends)
        in_progress.remove(name)

        # Merge child with parent
        merged = merge_metric(metric, parent)
        resolved[name] = merged
        return merged

    # Resolve all metrics
    for name in metrics:
        resolve(name)

    return resolved

File: sidemantic-rs/src/core/inheritance.rs
//! Model inheritance support
//!
//! Allows models to extend other models, inheriting dimensions, metrics,
//! relationships, segments, and pre-aggregations. Child values override parent values.

use std::collections::{HashMap, HashSet};

use crate::core::model::Model;
use crate::error::{Result, SidemanticError};

/// Merge a child model with its parent.
///
/// Child inherits all fields from parent, with child values taking precedence.
/// List fields (dimensions, metrics, relationships, segments, pre-aggregations) are merged by name,
/// with child items overriding parent items of the same name.
pub fn merge_model(child: &Model, parent: &Model) -> Model {
    // Start with parent's table/sql, override with child if set
    let table = child.table.clone().or_else(|| parent.table.clone());
    let sql = child.sql.clone().or_else(|| parent.sql.clone());
    let source_uri = child
        .source_uri
        .clone()
        .or_else(|| parent.source_uri.clone());
    let extends = child.extends.clone();
    let child_primary_key_columns = if child.primary_key_columns.is_empty() {
        vec![child.primary_key.clone()]
    } else {
        child.primary_key_columns.clone()
    };
    let parent_primary_key_columns = if parent.primary_key_columns.is_empty() {
        vec![parent.primary_key.clone()]
    } else {
        parent.primary_key_columns.clone()
    };
    let child_overrides_primary_key = child_primary_key_columns.len() > 1
        || child_primary_key_columns
            .first()
            .map(|value| value.as_str())
            != Some("id");
    let primary_key_columns = if child_overrides_primary_key {
        child_primary_key_columns
    } else {
        parent_primary_key_columns
    };
    let primary_key = primary_key_columns
        .first()
        .cloned()
        .unwrap_or_else(|| "id".to_string());
    let unique_keys = child
        .unique_keys
        .clone()
        .or_else(|| parent.unique_keys.clone());
    let description = child
        .description
        .clone()
        .or_else(|| parent.description.clone());
    let label = child.label.clone().or_else(|| parent.label.clone());
    let default_time_dimension = child
        .default_time_dimension
        .clone()
        .or_else(|| parent.default_time_dimension.clone());
    let default_grain = child
        .default_grain
        .clone()
        .or_else(|| parent.default_grain.clone());

    // Merge dimensions by name (child overrides parent)
    let mut dimensions_map: HashMap<String, _> = parent
        .dimensions
        .iter()
        .map(|d| (d.name.clone(), d.clone()))
        .collect();
    for dim in &child.dimensions {
        dimensions_map.insert(dim.name.clone(), dim.clone());
    }
    let dimensions: Vec<_> = dimensions_map.into_values().collect();

    // Merge metrics by name
    let mut metrics_map: HashMap<String, _> = parent
        .metrics
        .iter()
        .map(|m| (m.name.clone(), m.clone()))
        .collect();
    for metric in &child.metrics {
        metrics_map.insert(metric.name.clone(), metric.clone());
    }
    let metrics: Vec<_> = metrics_map.into_values().collect();

    // Merge relationships by name
    let mut relationships_map: HashMap<String, _> = parent
        .relationships
        .iter()
        .map(|r| (r.name.clone(), r.clone()))
        .collect();
    for rel in &child.relationships {
        relationships_map.insert(rel.name.clone(), rel.clone());
    }
    let relationships: Vec<_> = relationships_map.into_values().collect();

    // Merge segments by name
    let mut segments_map: HashMap<String, _> = parent
        .segments
        .iter()
        .map(|s| (s.name.clone(), s.clone()))
        .collect();
    for seg in &child.segments {
        segments_map.insert(seg.name.clone(), seg.clone());
    }
    let segments: Vec<_> = segments_map.into_values().collect();

    // Merge pre-aggregations by name
    let mut pre_aggs_map: HashMap<String, _> = parent
        .pre_aggregations
        .iter()
        .map(|p| (p.name.clone(), p.clone()))
        .collect();
    for pre_agg in &child.pre_aggregations {
        pre_aggs_map.insert(pre_agg.name.clone(), pre_agg.clone());
    }
    let pre_aggregations: Vec<_> = pre_aggs_map.into_values().collect();

    Model {
        name: child.name.clone(),
        table,
        sql,
        source_uri,
        extends,
        primary_key,
        primary_key_columns,
        unique_keys,
        dimensions,
        metrics,
        relationships,
        segments,
        pre_aggregations,
        default_time_dimension,
        default_grain,
        label,
        description,
    }
}

/// Resolve inheritance for all models.
///
/// Models with `extends` field are merged with their parent models.
/// Handles transitive inheritance (A extends B extends C).
/// Detects circular inheritance.
pub fn resolve_model_inheritance(
    models: HashMap<String, Model>,
    extends_map: &HashMap<String, String>,
) -> Result<HashMap<String, Model>> {
    let mut resolved: HashMap<String, Model> = HashMap::new();
    let mut in_progress: HashSet<String> = HashSet::new();

    fn resolve(
        name: &str,
        models: &HashMap<String, Model>,
        extends_map: &HashMap<String, String>,
        resolved: &mut HashMap<String, Model>,
        in_progress: &mut HashSet<String>,
    ) -> Result<Model> {
        // Already resolved
        if let Some(model) = resolved.get(name) {
            return Ok(model.clone());
        }

        // Check for circular inheritance
        if in_progress.contains(name) {
            return Err(SidemanticError::Validation(format!(
                "Circular inheritance detected for model '{name}'"
            )));
        }

        // Get the model
        let model = models
            .get(name)
            .ok_or_else(|| SidemanticError::Validation(format!("Model '{name}' not found")))?;

        // If no inheritance, just return as-is
        let parent_name = match extends_map.get(name) {
            Some(parent) => parent,
            None => {
                resolved.insert(name.to_string(), model.clone());
                return Ok(model.clone());
            }
        };

        // Resolve parent first
        in_progress.insert(name.to_string());
        let parent = resolve(parent_name, models, extends_map, resolved, in_progress)?;
        in_progress.remove(name);

        // Merge child with parent
        let merged = merge_model(model, &parent);
        resolved.insert(name.to_string(), merged.clone());
        Ok(merged)
    }

    // Resolve all models
    let names: Vec<_> = models.keys().cloned().collect();
    for name in names {
        resolve(&name, &models, extends_map, &mut resolved, &mut in_progress)?;
    }

    Ok(resolved)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::{Dimension, Metric};

    #[test]
    fn test_merge_model_basic() {
        let parent = Model::new("base", "id")
            .with_table("base_table")
            .with_dimension(Dimension::categorical("status"))
            .with_metric(Metric::sum("revenue", "amount"));

        let child = Model::new("child", "id")
            .with_dimension(Dimension::categorical("category"))
            .with_metric(Metric::count("order_count"));

        let merged = merge_model(&child, &parent);

        assert_eq!(merged.name, "child");
        assert_eq!(merged.table, Some("base_table".to_string()));
        assert_eq!(merged.dimensions.len(), 2); // status + category
        assert_eq!(merged.metrics.len(), 2); // revenue + order_count
    }

    #[test]
    fn test_merge_model_override() {
        let parent = Model::new("base", "id")
            .with_table("base_table")
            .with_metric(Metric::sum("revenue", "amount"));

        let child = Model::new("child", "id")
            .with_table("child_table")
            .with_metric(Metric::sum("revenue", "total_amount")); // Override

        let merged = merge_model(&child, &parent);

        assert_eq!(merged.table, Some("child_table".to_string()));
        // Child's metric should override parent's
        let revenue = merged.metrics.iter().find(|m| m.name == "revenue").unwrap();
        assert_eq!(revenue.sql, Some("total_amount".to_string()));
    }

    #[test]
    fn test_resolve_inheritance() {
        let mut models = HashMap::new();
        models.insert(
            "base".to_string(),
            Model::new("base", "id")
                .with_table("base_table")
                .with_dimension(Dimension::categorical("status")),
        );
        models.insert(
            "child".to_string(),
            Model::new("child", "id").with_metric(Metric::count("order_count")),
        );

        let mut extends_map = HashMap::new();
        extends_map.insert("child".to_string(), "base".to_string());

        let resolved = resolve_model_inheritance(models, &extends_map).unwrap();

        let child = resolved.get("child").unwrap();
        assert_eq!(child.table, Some("base_table".to_string()));
        assert_eq!(child.dimensions.len(), 1);
        assert_eq!(child.metrics.len(), 1);
    }

    #[test]
    fn test_circular_inheritance_detected() {
        let mut models = HashMap::new();
        models.insert("a".to_string(), Model::new("a", "id"));
        models.insert("b".to_string(), Model::new("b", "id"));

        let mut extends_map = HashMap::new();
        extends_map.insert("a".to_string(), "b".to_string());
        extends_map.insert("b".to_string(), "a".to_string());

        let result = resolve_model_inheritance(models, &extends_map);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("Circular"));
    }
}
